src/ripple/rpc/handlers/WalletPropose.cpp

```diff
@@ -31,9 +31,30 @@
 #include <ripple/rpc/handlers/WalletPropose.h>
 #include <ed25519-donna/ed25519.h>
 #include <boost/optional.hpp>
+#include <cmath>
+#include <map>
 
 namespace ripple {
 
+double
+shannon_entropy (std::string const& input)
+{
+    std::map<int, double> freq;
+
+    for (auto const& c : input)
+        freq[c]++;
+
+    double se = 0;
+
+    for (auto& f : freq)
+        f.second /= input.length();
+
+    for (auto const& f : freq)
+        se += (f.second) * log2(f.second);
+
+    return std::floor (-se * input.length());
```

Comment:
The function might be a little misnamed. Here's the rationale:

I'm trying to figure out how many bits of entropy this password contains. I do this by taking the number of bits per character in the ideal encoding (`-se`) and then multiplying it by the length of the password. This gives us back the number of bits this password could represent in if it used an ideal encoding.

Since I can only have a integral number of bits, I can't deal with something like 123.875348 bits. Additionally, since I'm using the result as an estimate on the entropy, I prefer to be ultra conservative (that is, to round down). This means that, in the worst case, I simply underestimate the entropy and issue a warning for a strong passphrase as opposed to not warning for a passphrase on the "cusp" of what I've defined as strong and weak.